src/core/desk.sol

```diff
@@ -0,0 +1,132 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.4.24;
+
+import "../../lib/dss-add-ilk-spell/lib/dss-deploy/lib/dss/src/lib.sol";
+
+contract PileLike {
+    function want() public returns (int);
+}
+
+contract DistributorLike {
+    function waterfallRepay() public;
+    function makerRepay() public;
+}
+
+contract OperatorLike {
+    function borrow(address, uint) public;
+    function repay(address, uint) public;
+    function balance() public returns (uint);
+    function file(bytes32, bool) public;
+
+    bool public supplyActive;
+    bool public redeemActive;
+}
+
+// Desk
+// Keeps track of the tranches. Manages the interfacing between the tranche side and borrower side of the contracts.
+contract Desk is DSNote {
+
+    // --- Auth ---
+    mapping (address => uint) public wards;
+    function rely(address usr) public auth note { wards[usr] = 1; }
+    function deny(address usr) public auth note { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    // --- Tranches ---
+
+    struct Tranche {
+        uint ratio;
+        OperatorLike operator;
+    }
+
+    Tranche[] tranches;
+
+    // --- Data ---
+    PileLike public pile;
+    DistributorLike public distributor;
+
+    bool public flowThrough;
+
+    constructor (address pile_, address distributor_) public {
+        wards[msg.sender] = 1;
+        pile = PileLike(pile_);
+        distributor = DistributorLike(distributor_);
+        flowThrough = false;
+    }
+
+    function depend (bytes32 what, address addr) public auth {
+        if (what == "pile") { pile = PileLike(addr); }
+        else if (what == "distributor") { distributor = DistributorLike(addr); }
+        else revert();
+    }
+
+    function file(bytes32 what, bool data) public auth {
+        if (what == "flowThrough") { flowThrough = data; }
+    }
+
+    function file(uint i, bytes32 what, bool data) public auth {
+        tranches[i].operator.file(what, data);
+    }
+
+    // --- Calls ---
+
+    function addTranche(uint ratio, address operator_) public auth {
+        Tranche memory t;
+        t.ratio = ratio;
+        t.operator = OperatorLike(operator_);
+        tranches.push(t);
+    }
+
+    function returnOperator(uint i) public auth returns (address){
+       return address(tranches[i].operator);
+    }
+
+    function returnRatio(uint i) public auth returns (uint) {
+        return tranches[i].ratio;
+    }
+
+    function balance(uint i) public auth {
+
+        // balance methods should:
+        // iterate through the tranches
+        // quant debt =  how much is
+
+        Tranche memory t = tranches[i];
+        // if capital should flow through, all funds in reserve should be moved in pile
+        if (flowThrough) {
+            if (t.operator.supplyActive()) {
```

Comment:
supplyActive just controls if you can invest/put money in the tranche reserve, not if you can borrow from it. supply can be deactivated but money still can be borrowed from the reserve